ENH: implement __sizeof__ for estimators

stsievert opened the pull request:
#### Reference Issues/PRs
Closes #8642 

#### What does this implement/fix? Explain your changes.
It implements `__sizeof__` for Estimators as mentioned in https://github.com/scikit-learn/scikit-learn/issues/8642#issue-216883385.

TODO:

- [x] implement and test `__sizeof__`
- [ ] implement `__sizeof__` for trees

stsievert committed:
ENH: implement sizeof

stsievert committed:
TST: include pickle message size in sizeof tests

stsievert commented:
This PR is not exact, and doesn't not count any overhead. In the tests, it's fairly straightforward to get within ±5%, at least for decently large estimators. This is mentioned more by @mrocklin in https://github.com/scikit-learn/scikit-learn/issues/8642#issuecomment-395892938 and https://github.com/scikit-learn/scikit-learn/issues/8642#issuecomment-397004343.

rth reviewed (commented):
Thanks @stsivert. It would be useful to add this to common tests https://github.com/scikit-learn/scikit-learn/issues/8642#issuecomment-394023681 which can be added e.g. to `sklearn.utils.estimator_checks::check_estimator_pickle`. 

It may be fine if for some estimaors the size of the pickle is not equal with some tolerance to `__sizeof__` result, but at least it would be good to know which ones.

stsievert committed:
Add more general test

stsievert committed:
Check only for large sizes

stsievert committed:
pytest approx

stsievert committed:
temp pytest edit

stsievert committed:
MAINT: weaken test; not many feautres

stsievert committed:
better test

stsievert committed:
Test with more features

stsievert committed:
Add large features to check_estimators

stsievert committed:
Add sizeof to more estimators

Modifying *Mixin classes is necessary because many classes
override BaseEstimator's implementation with (for example)

    class Birch(ClusterMixin, TransformerMixin, BaseEstimator):

This means that Birch will use ClusterMixin's implementation of
__sizeof__ and override BaseEstimator's

stsievert committed:
Remove unused import

stsievert committed:
small testing modification

jnothman reviewed (commented):
I thought we decided we only need to implement it for C/++ data structures since the caller should be dealing with estimation for plain old objects.

stsievert committed:
more info on tests

stsievert committed:
exact recursive strategy for test verification

stsievert committed:
lint

cmarmo commented:
Hi @stsievert, I'm checking the status of PRs and issues in the [0.23 milestone](milestone/35).
Do you think you could find some time to work on this one? Thanks a lot for your time!

stsievert committed:
Remove verbose flag

rth commented on sklearn/base.py:
```diff
@@ -332,10 +336,19 @@ def _get_tags(self):
     _estimator_type = "classifier"
 
+    def __sizeof__(self):
+        return _sizeof(self)
```
Why do we need it in each mixin, wouldn't just `BaseEstimator` be enough? Everything inherits from it.

stsievert replied:
Do all estimators inherit from `BaseEstimator`? It doesn't look like it in the code because `class ClassifierMixin:` is used, not `class ClassifierMixin(BaseEstimator):`. It also doesn't look like it in Python:

``` python
>>> from sklearn.base import ClassifierMixin
>>> ClassifierMixin.mro()
[sklearn.base.ClassifierMixin, object]
```

rth replied:
> Do all estimators inherit from BaseEstimator

Estimators - yes. There might be other objects that don't though. E.g. losses (https://github.com/scikit-learn/scikit-learn/blob/557218c021d58c7455b26820f331a9c388c9cb4d/sklearn/_loss/glm_distribution.py#L20)

`ClassifierMixin` is a mixin, so it's never going to be used alone. It's always BaseEstimator + e.g. ClassifierMixin.

So there is no need to add this method to Mixins, you can be sure they are always used in combination with BaseEstimator. However this might need to be done for other custom objects.

rth commented on sklearn/tests/test_base.py:
```diff
@@ -526,3 +530,21 @@ def fit(self, X, y=None):
+    init = sys.getsizeof(est)
+    assert init <= 200
+
+    est.fit(*make_classification(n_features=1000))
```
```suggestion
    X, y = make_classification(n_features=1000)
    est.fit(X, y)
```

rth commented on sklearn/base.py:
```diff
@@ -332,10 +336,19 @@ def _get_tags(self):
 
+def _sizeof(cls):
+    #  return sum(x.nbytes if hasattr(x, 'nbytes') else sys.getsizeof(x)
+    #  for x in cls.__dict__.values())
```
Does the previous version not work?
Or is `sys.getsizeof` using x.nbytes when available? I would have imagined a specialized handling maybe for ndarray, csr/csc arrays and DataFrames ?

stsievert replied:
I've uncommented this version (and added another check for `not callable(x.nbytes)`). `x.nbytes` is a better approximation than `sys.getsizeof`:

``` python
>>> N = 1_000_000
>>> x = np.arange(N).reshape(-1, N)
>>> x.shape
(1, 1000000)
>>> sys.getsizeof(x)
112
>>> x.nbytes
8000000
>>> sys.getsizeof(x.flatten())
8000096
```

rth reviewed (commented):
> I thought we decided we only need to implement it for C/++ data structures since the caller should be dealing with estimation for plain old objects.

I would be OK with it in BaseEstimator, and any Cython/C objects that need it. Having a simple way to check the size of an estimator won't hurt even outside of Dask.

Please ping once CI is passing.

rth commented:
(Also it might be good to sync with master)

stsievert committed:
Update sklearn/tests/test_base.py

Co-Authored-By: Roman Yurchak <rth.yurchak@gmail.com>

stsievert committed:
More accurate sizeof

stsievert committed:
Merge branch 'master' into sizeof

stsievert committed:
Merge branch 'sizeof' of https://github.com/stsievert/scikit-learn into sizeof

rth commented on sklearn/tests/test_base.py:
```diff
@@ -511,3 +515,22 @@ def fit(self, X, y=None):
+
+    # Does it take __sizeof__ byte to encode approximately?
+    msg = pickle.dumps(est)
+    assert len(msg) == pytest.approx(sys.getsizeof(est), rel=0.05)
```
Maybe increase the tolerance a bit, even 20% is not so bad ?

jnothman commented:
So it's okay that this can double count objects with shared references?

stsievert commented:
> So it's okay that this can double count objects with shared references?

Yeah, I think it'd be best to forget about looping over `__dict__` in `__sizeof__`. Instead, how 'bout this?

``` python
class BaseEstimator:

    @property
    def nbytes(self):
        return sum(x.nbytes if hasattr(x, "nbytes") else sys.getsizeof(x) for x in self.__dict__.values())

    ...
```

That'd resolve @jnothman's concerns and provide an accurate estimate of an estimator's memory usage.

stsievert committed:
MAINT: __sizeof__ => nbytes
